transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,706 @@
+            _ => Poll::Pending,
+        }
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let sock_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if sock_addr.port() == 0 || sock_addr.ip().is_unspecified() {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+
+        let remote = addr.clone(); // used for logging
+        trace!("dialing addr={}", remote);
+
+        let config = self.config.clone();
+        let our_fingerprint = self.cert_fingerprint();
+        let id_keys = self.id_keys.clone();
+
+        let first_listener = self
+            .listeners
+            .iter()
+            .next()
+            .ok_or(TransportError::Other(Error::NoListeners))?;
+        let udp_mux = first_listener.udp_mux.clone();
+
+        // [`Transport::dial`] should do no work unless the returned [`Future`] is polled. Thus
+        // do the `set_remote_description` call within the [`Future`].
+        Ok(async move {
+            let remote_fingerprint = fingerprint_from_addr(&addr)
+                .map(|f| Fingerprint::from(f.iter()))
+                .ok_or(Error::InvalidMultiaddr(addr.clone()))?;
+
+            let conn = WebRTCConnection::connect(
+                sock_addr,
+                config.into_inner(),
+                udp_mux,
+                &remote_fingerprint,
+            )
+            .await?;
+
+            // Open a data channel to do Noise on top and verify the remote.
+            let data_channel = conn.create_initial_upgrade_data_channel(None).await?;
+
+            trace!("noise handshake with addr={}", remote);
+            let peer_id = perform_noise_handshake(
+                id_keys,
+                PollDataChannel::new(data_channel.clone()),
+                &our_fingerprint,
+                remote_fingerprint.as_ref(),
+            )
+            .await?;
+
+            trace!("verifying peer's identity addr={}", remote);
+            let peer_id_from_addr = PeerId::try_from_multiaddr(&addr);
+            if peer_id_from_addr.is_none() || peer_id_from_addr.unwrap() != peer_id {
+                return Err(Error::InvalidPeerID {
+                    expected: peer_id_from_addr,
+                    got: peer_id,
+                });
+            }
+
+            // Close the initial data channel after noise handshake is done.
+            data_channel
+                .close()
+                .await
+                .map_err(|e| Error::WebRTC(webrtc::Error::Data(e)))?;
+
+            let mut c = Connection::new(conn.into_inner()).await;
+            // TODO: default buffer size is too small to fit some messages. Possibly remove once
+            // https://github.com/webrtc-rs/sctp/issues/28 is fixed.
+            c.set_data_channels_read_buf_capacity(8192 * 10);
+            Ok((peer_id, c))
```

Comment:
because our type does not provide all of the methods of https://docs.rs/webrtc/0.4.0/webrtc/peer_connection/struct.RTCPeerConnection.html ?

Reply:
Would it be feasible to create the necessary, potentially more high-level APIs that we need on our type?

Reply:
I just realised that we actually own `Connection`. I thought that was a webrtc type. I think it would make sense to merge `Connection` and `WebRTCConnection` so we don't have to convert here. What do you think?

Reply:
I don't see a reason why we need to drag `WebRTCConnection` into `Connection`. Maybe we should just rename `WebRTCConnection` to `WebRTCConnectionBuilder` (or smth like that) because this is what it does actually.

Reply:
They both represent the connection but at different stages of its life-cycle actually. We could express this with a generic life-cycle parameter and attach certain functions only to the corresponding lifecycle stage.

Something like:

```rust
struct Connection<TStage> {
	phantom: PhantomData<TStage>
}

impl Connection<Initial> {
    async fn accept() -> Connection<Authenticating> {}
	async fn connect() -> Connection<Authenticating> {}
}

impl Connection<Authenticating> {
	async fn create_authentication_channel(&self) -> Result<Arc<DataChannel>> {}
    /// Attempts to complete the upgrade of this connection with the resolved PeerId from the noise handshake.
    ////
    //// This function fails in case the provided PeerId does not match the one we are expecting from the address.
    async fn complete_uprade(self, resolved_peer_id: PeerId) -> Result<Connection<Ready>, Error> { }
}
```

Instead of a `PhantomData`, we could also store `TStage` directly in case we need to associate some data that we only need for a certain stage. We could even embed the role in the in there based on whether `accept` or `connect` was called which would allow us to internally do the right thing within `create_authentication_channel` without passing a parameter, i.e.

```rust
impl Connection<Initial> {
    async fn accept() -> Connection<AuthenticatingListener> {}
	async fn connect() -> Connection<AuthenticatingDialer> {}
}

impl Connection<AuthenticatingListener> {
	async fn create_authentication_channel(&self) -> Result<Arc<DataChannel>> {}
}

impl Connection<AuthenticatingDialer> {
	async fn create_authentication_channel(&self) -> Result<Arc<DataChannel>> {}
}
```

Let me know what you think of the idea. I would nicely express the connection upgrade process in a type-safe way which also acts as documentation.